An error occurred between opentelemetry modules

<!--

Welcome to ingress-nginx!  For a smooth issue process, try to answer the following questions.
Don't worry if they're not all applicable; just try to include what you can :-)

If you need to include code snippets or logs, please put them in fenced code
blocks.  If they're super-long, please use the details tag like
<details><summary>super-long log</summary> lots of stuff </details>

-->

<!--

IMPORTANT!!!

Please complete the next sections or the issue will be closed.
This questions are the first thing we need to know to understand the context.

-->

**NGINX Ingress controller version** (exec into the pod and run nginx-ingress-controller --version.):

```
bash-5.1$ ./nginx-ingress-controller --version
-------------------------------------------------------------------------------
NGINX Ingress controller
  Release:       v1.1.3
  Build:         9d3a285f19a704524439c75b947e2189406565ab
  Repository:    https://github.com/kubernetes/ingress-nginx
  nginx version: nginx/1.19.10

-------------------------------------------------------------------------------

[root@trace1 ingress-nginx]# kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.8", GitCommit:"7061dbbf75f9f82e8ab21f9be7e8ffcaae8e0d44", GitTreeState:"clean", BuildDate:"2022-03-16T14:10:06Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.8", GitCommit:"7061dbbf75f9f82e8ab21f9be7e8ffcaae8e0d44", GitTreeState:"clean", BuildDate:"2022-03-16T14:04:34Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}

```

**Environment**:

- **Cloud provider or hardware configuration**:
- baremetal, kubernetes on centos7 
```
[root@trace1 ingress-nginx]# cat /etc/*release
CentOS Linux release 7.9.2009 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.9.2009 (Core)
CentOS Linux release 7.9.2009 (Core)
```

- **Cloud provider or hardware configuration**:

- **Kernel** (e.g. `uname -a`):

```
  [root@trace1 ingress-nginx]# uname -a
Linux trace1 3.10.0-1160.59.1.el7.x86_64 #1 SMP Wed Feb 23 16:47:03 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```
- **Install tools**:
  - `Please mention how/where was the cluster created like kubeadm/kops/minikube/kind etc. `
  - kubespray 2.18 
  - helm chart   

- **Basic cluster related info**:
  - `kubectl get nodes -o wide`
```
  [root@trace1 ingress-nginx]# kubectl get nodes
NAME     STATUS   ROLES                  AGE   VERSION
trace1   Ready    control-plane,master   3d    v1.22.8
trace2   Ready    <none>                 3d    v1.22.8
```

- **How was the ingress-nginx-controller installed**:
  - If helm was used then please show output of `helm ls -A | grep -i ingress`
```
git clone https://github.com/kubernetes/ingress-nginx.git
cd ingress-nginx/charts/ingress-nginx
cat << EOF >> extra.yml 
  extraModules:
  - name: opentelemetry
    image: gcr.io/k8s-staging-ingress-nginx/opentelemetry:v20220331-controller-v1.1.2-36-g7517b7ecf
EOF

helm install ingress-nginx . -f extra.yml`


[root@trace1 ingress-nginx]# helm ls -A
NAME            NAMESPACE       REVISION        UPDATED                                 STATUS          CHART                  APP VERSION
ingress-nginx   default         1               2022-04-06 05:03:26.400791535 +0000 UTC deployed        ingress-nginx-4.0.19   1.1.3
```

  - If helm was used then please show output of `helm -n <ingresscontrollernamepspace> get values <helmreleasename>`
```
[root@trace1 ingress-nginx]# helm get values ingress-nginx
USER-SUPPLIED VALUES:
null
```
  - If helm was not used, then copy/paste the complete precise command used to install the controller, along with the flags and options used
  - if you have more than one instance of the ingress-nginx-controller installed in the same cluster, please provide details for all the instances

- **Current State of the controller**:
  - `kubectl describe ingressclasses`
  - `kubectl -n <ingresscontrollernamespace> get all -A -o wide`
  - `kubectl -n <ingresscontrollernamespace> describe po <ingresscontrollerpodname>`
```
  [root@trace1 ingress-nginx]# kubectl get po -n default
NAME                                        READY   STATUS    RESTARTS   AGE
ingress-nginx-controller-5c5846d6f7-2b7gr   1/1     Running   0          20m

[root@trace1 ingress-nginx]# kubectl describe po ingress-nginx-controller-5c5846d6f7-2b7gr 
Name:         ingress-nginx-controller-5c5846d6f7-2b7gr
Namespace:    default
Priority:     0
Node:         trace2/192.168.16.69
Start Time:   Wed, 06 Apr 2022 05:03:28 +0000
Labels:       app.kubernetes.io/component=controller
              app.kubernetes.io/instance=ingress-nginx
              app.kubernetes.io/name=ingress-nginx
              pod-template-hash=5c5846d6f7
Annotations:  cni.projectcalico.org/containerID: 6446671f8ac137a1024148588551902e25e715c3a22f35cbed1544a9a096682a
              cni.projectcalico.org/podIP: 10.233.94.74/32
              cni.projectcalico.org/podIPs: 10.233.94.74/32
Status:       Running
IP:           10.233.94.74
IPs:
  IP:           10.233.94.74
Controlled By:  ReplicaSet/ingress-nginx-controller-5c5846d6f7
Init Containers:
  opentelemetry:
    Container ID:  containerd://24b7aa7186e49c44cfa5ba59f7a580ac4e61f9f527e2052176d927196559b752
    Image:         yjkim1ntels/ingress-nginx:opentelemetry
    Image ID:      docker.io/yjkim1ntels/ingress-nginx@sha256:d86b3679ad13c510d5991f5e2c6e34dc5d2e957e60b493487cff875a138ed806
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
      /usr/local/bin/init_module.sh
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 06 Apr 2022 05:03:34 +0000
      Finished:     Wed, 06 Apr 2022 05:03:34 +0000
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /modules_mount from modules (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-xn5cv (ro)
Containers:
  controller:
    Container ID:  containerd://96113bb6bdbd35d91534ef40c5794d2ac1c551b06b64bd0d91dbb57c4042453f
    Image:         k8s.gcr.io/ingress-nginx/controller:v1.1.3@sha256:31f47c1e202b39fadecf822a9b76370bd4baed199a005b3e7d4d1455f4fd3fe2
    Image ID:      k8s.gcr.io/ingress-nginx/controller@sha256:31f47c1e202b39fadecf822a9b76370bd4baed199a005b3e7d4d1455f4fd3fe2
    Ports:         80/TCP, 443/TCP, 8443/TCP
    Host Ports:    0/TCP, 0/TCP, 0/TCP
    Args:
      /nginx-ingress-controller
      --publish-service=$(POD_NAMESPACE)/ingress-nginx-controller
      --election-id=ingress-controller-leader
      --controller-class=k8s.io/ingress-nginx
      --ingress-class=nginx
      --configmap=$(POD_NAMESPACE)/ingress-nginx-controller
      --validating-webhook=:8443
      --validating-webhook-certificate=/usr/local/certificates/cert
      --validating-webhook-key=/usr/local/certificates/key
    State:          Running
      Started:      Wed, 06 Apr 2022 05:03:35 +0000
    Ready:          True
    Restart Count:  0
    Requests:
      cpu:      100m
      memory:   90Mi
    Liveness:   http-get http://:10254/healthz delay=10s timeout=1s period=10s #success=1 #failure=5
    Readiness:  http-get http://:10254/healthz delay=10s timeout=1s period=10s #success=1 #failure=3
    Environment:
      POD_NAME:       ingress-nginx-controller-5c5846d6f7-2b7gr (v1:metadata.name)
      POD_NAMESPACE:  default (v1:metadata.namespace)
      LD_PRELOAD:     /usr/local/lib/libmimalloc.so
    Mounts:
      /modules_mount from modules (rw)
      /usr/local/certificates/ from webhook-cert (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-xn5cv (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True
Volumes:
  modules:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  webhook-cert:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  ingress-nginx-admission
    Optional:    false
  kube-api-access-xn5cv:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              kubernetes.io/os=linux
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason     Age   From                      Message
  ----    ------     ----  ----                      -------
  Normal  Scheduled  20m   default-scheduler         Successfully assigned default/ingress-nginx-controller-5c5846d6f7-2b7gr to trace2
  Normal  Pulling    20m   kubelet                   Pulling image "yjkim1ntels/ingress-nginx:opentelemetry"
  Normal  Pulled     20m   kubelet                   Successfully pulled image "yjkim1ntels/ingress-nginx:opentelemetry" in 5.495976532s
  Normal  Created    20m   kubelet                   Created container opentelemetry
  Normal  Started    20m   kubelet                   Started container opentelemetry
  Normal  Pulled     20m   kubelet                   Container image "k8s.gcr.io/ingress-nginx/controller:v1.1.3@sha256:31f47c1e202b39fadecf822a9b76370bd4baed199a005b3e7d4d1455f4fd3fe2" already present on machine
  Normal  Created    20m   kubelet                   Created container controller
  Normal  Started    20m   kubelet                   Started container controller
  Normal  RELOAD     20m   nginx-ingress-controller  NGINX reload triggered due to a change in configuration
```



**What happened**:

<!-- (please include exact error messages if you can) -->

* An error occurred while adding the openteletry module as a sidecar to ingress-nginx.

```
ls -al /modules_mount/etc/nginx/modules/modules/otel_ngx_module.so
-rwxr-xr-x    1 root     root       8077416 Apr  6 04:20 /modules_mount/etc/nginx/modules/modules/otel_ngx_module.so

# add nginx.conf load_module 

bash-5.1$ cat nginx.conf | head -n 3
load_module /modules_mount/etc/nginx/modules/modules/otel_ngx_module.so;
# Configuration checksum: 6642706386070326205

nginx -s reload
 
# error 
2022/04/06 04:45:48 [emerg] 619#619: dlopen() "/modules_mount/etc/nginx/modules/modules/otel_ngx_module.so" failed (Error relocating /modules_mount/etc/nginx/modules/modules/otel_ngx_module.so: _ZN13opentelemetry5proto5trace2v14Span8CopyFromERKS3_: symbol not found) in /etc/nginx/nginx.conf:1
nginx: [emerg] dlopen() "/modules_mount/etc/nginx/modules/modules/otel_ngx_module.so" failed (Error relocating /modules_mount/etc/nginx/modules/modules/otel_ngx_module.so: _ZN13opentelemetry5proto5trace2v14Span8CopyFromERKS3_: symbol not found) in /etc/nginx/nginx.conf:1
```

**What you expected to happen**:

- I want the nginx-ingress-controller and opentelemetry module to work together.

@dmathieu  
I got a err message 

```

/tmp/build # cd nginx-1.19.10/
/tmp/build/nginx-1.19.10 # ls
CHANGES     LICENSE     README      conf        contrib     man         src
CHANGES.ru  Makefile    auto        configure   html        objs
/tmp/build/nginx-1.19.10 # ./configure '--prefix=/usr/local/nginx' --with-compat '--add-dynamic-module=/tmp/build/opent
elemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx'
checking for OS
 + Linux 3.10.0-1160.59.1.el7.x86_64 x86_64
checking for C compiler ... found
 + using GNU C compiler
 + gcc version: 10.3.1 20210424 (Alpine 10.3.1_git20210424) 
checking for gcc -pipe switch ... found
checking for -Wl,-E switch ... found
checking for gcc builtin atomic operations ... found
checking for C99 variadic macros ... found
checking for gcc variadic macros ... found
checking for gcc builtin 64 bit byteswap ... found
checking for unistd.h ... found
checking for inttypes.h ... found
checking for limits.h ... found
checking for sys/filio.h ... not found
checking for sys/param.h ... found
checking for sys/mount.h ... found
checking for sys/statvfs.h ... found
checking for crypt.h ... found
checking for Linux specific features
checking for epoll ... found
checking for EPOLLRDHUP ... found
checking for EPOLLEXCLUSIVE ... found
checking for eventfd() ... found
checking for O_PATH ... found
checking for sendfile() ... found
checking for sendfile64() ... found
checking for sys/prctl.h ... found
checking for prctl(PR_SET_DUMPABLE) ... found
checking for prctl(PR_SET_KEEPCAPS) ... found
checking for capabilities ... not found
checking for crypt_r() ... found
checking for sys/vfs.h ... found
checking for nobody group ... found
checking for poll() ... found
checking for /dev/poll ... not found
checking for kqueue ... not found
checking for crypt() ... found
checking for F_READAHEAD ... not found
checking for posix_fadvise() ... found
checking for O_DIRECT ... found
checking for F_NOCACHE ... not found
checking for directio() ... not found
checking for statfs() ... found
checking for statvfs() ... found
checking for dlopen() ... found
checking for sched_yield() ... found
checking for sched_setaffinity() ... found
checking for SO_SETFIB ... not found
checking for SO_REUSEPORT ... found
checking for SO_ACCEPTFILTER ... not found
checking for SO_BINDANY ... not found
checking for IP_TRANSPARENT ... found
checking for IP_BINDANY ... not found
checking for IP_BIND_ADDRESS_NO_PORT ... found
checking for IP_RECVDSTADDR ... not found
checking for IP_SENDSRCADDR ... not found
checking for IP_PKTINFO ... found
checking for IPV6_RECVPKTINFO ... found
checking for TCP_DEFER_ACCEPT ... found
checking for TCP_KEEPIDLE ... found
checking for TCP_FASTOPEN ... found
checking for TCP_INFO ... found
checking for accept4() ... found
checking for int size ... 4 bytes
checking for long size ... 8 bytes
checking for long long size ... 8 bytes
checking for void * size ... 8 bytes
checking for uint32_t ... found
checking for uint64_t ... found
checking for sig_atomic_t ... found
checking for sig_atomic_t size ... 4 bytes
checking for socklen_t ... found
checking for in_addr_t ... found
checking for in_port_t ... found
checking for rlim_t ... found
checking for uintptr_t ... uintptr_t found
checking for system byte ordering ... little endian
checking for size_t size ... 8 bytes
checking for off_t size ... 8 bytes
checking for time_t size ... 8 bytes
checking for AF_INET6 ... found
checking for setproctitle() ... not found
checking for pread() ... found
checking for pwrite() ... found
checking for pwritev() ... found
checking for strerrordesc_np() ... not found
checking for sys_nerr ... not found
checking for _sys_nerr ... not found
checking for localtime_r() ... found
checking for clock_gettime(CLOCK_MONOTONIC) ... found
checking for posix_memalign() ... found
checking for memalign() ... found
checking for mmap(MAP_ANON|MAP_SHARED) ... found
checking for mmap("/dev/zero", MAP_SHARED) ... found
checking for System V shared memory ... found
checking for POSIX semaphores ... found
checking for struct msghdr.msg_control ... found
checking for ioctl(FIONBIO) ... found
checking for ioctl(FIONREAD) ... found
checking for struct tm.tm_gmtoff ... found
checking for struct dirent.d_namlen ... not found
checking for struct dirent.d_type ... found
checking for sysconf(_SC_NPROCESSORS_ONLN) ... found
checking for sysconf(_SC_LEVEL1_DCACHE_LINESIZE) ... not found
checking for openat(), fstatat() ... found
checking for getaddrinfo() ... found
configuring additional dynamic modules
adding module in /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx
 + otel_ngx_module was configured
checking for PCRE library ... found
checking for PCRE JIT support ... found
checking for zlib library ... found
creating objs/Makefile

Configuration summary
  + using system PCRE library
  + OpenSSL library is not used
  + using system zlib library

  nginx path prefix: "/usr/local/nginx"
  nginx binary file: "/usr/local/nginx/sbin/nginx"
  nginx modules path: "/usr/local/nginx/modules"
  nginx configuration prefix: "/usr/local/nginx/conf"
  nginx configuration file: "/usr/local/nginx/conf/nginx.conf"
  nginx pid file: "/usr/local/nginx/logs/nginx.pid"
  nginx error log file: "/usr/local/nginx/logs/error.log"
  nginx http access log file: "/usr/local/nginx/logs/access.log"
  nginx http client request body temporary files: "client_body_temp"
  nginx http proxy temporary files: "proxy_temp"
  nginx http fastcgi temporary files: "fastcgi_temp"
  nginx http uwsgi temporary files: "uwsgi_temp"
  nginx http scgi temporary files: "scgi_temp"

/tmp/build/nginx-1.19.10 # ls -al /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrum
entation/nginx/
CMakeLists.txt  README.md       ci/             config          nginx.cmake     src/            test/

/tmp/build/nginx-1.19.10 # ls -al /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrum
entation/nginx/
total 24
drwxrwxr-x    5 root     root           111 Apr  6 07:07 .
drwxrwxr-x    5 root     root            61 Apr  6 07:07 ..
-rw-rw-r--    1 root     root          1019 Mar 29 04:20 CMakeLists.txt
-rw-rw-r--    1 root     root          8090 Mar 29 04:20 README.md
drwxrwxr-x    2 root     root            34 Apr  6 07:07 ci
-rw-rw-r--    1 root     root           908 Mar 29 04:20 config
-rw-rw-r--    1 root     root          1170 Mar 29 04:20 nginx.cmake
drwxrwxr-x    2 root     root          4096 Apr  6 07:07 src
drwxrwxr-x    6 root     root           112 Apr  6 07:07 test

/tmp/build/nginx-1.19.10 # make modules
make -f objs/Makefile modules
make[1]: Entering directory '/tmp/build/nginx-1.19.10'
cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
        -o objs/addon/src/agent_config.o \
        /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/agent_config.cpp
cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
        -o objs/addon/src/nginx_config.o \
        /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/nginx_config.cpp
cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
        -o objs/addon/src/otel_ngx_module.o \
        /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/otel_ngx_module.cpp
cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
        -o objs/addon/src/propagate.o \
        /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/propagate.cpp
cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
        -o objs/addon/src/script.o \
        /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/script.cpp
cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
        -o objs/addon/src/toml.o \
        /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/toml.c
cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
        -o objs/addon/src/trace_context.o \
        /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/trace_context.cpp
cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
        -o objs/otel_ngx_module_modules.o \
        objs/otel_ngx_module_modules.c
cc -o objs/otel_ngx_module.so \
objs/addon/src/agent_config.o \
objs/addon/src/nginx_config.o \
objs/addon/src/otel_ngx_module.o \
objs/addon/src/propagate.o \
objs/addon/src/script.o \
objs/addon/src/toml.o \
objs/addon/src/trace_context.o \
objs/otel_ngx_module_modules.o \
-lstdc++ -lopentelemetry_common -lopentelemetry_resources -lopentelemetry_trace -lopentelemetry_exporter_otlp_grpc -lopentelemetry_otlp_recordable \
-shared
/usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lopentelemetry_common
/usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lopentelemetry_resources
/usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lopentelemetry_trace
/usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lopentelemetry_exporter_otlp_grpc
/usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lopentelemetry_otlp_recordable
collect2: error: ld returned 1 exit status
make[1]: *** [objs/Makefile:1210: objs/otel_ngx_module.so] Error 1
make[1]: Leaving directory '/tmp/build/nginx-1.19.10'
make: *** [Makefile:16: modules] Error 2
```

Is this a related error?

We can edit the Dockerfole for opentelemetry under /images and promote
again. What would be a good way to get a list of missing modules.

Thanks,
; Long

On Wed, 6 Apr, 2022, 1:21 PM oct28-yjkim, ***@***.***> wrote:

> @dmathieu <https://github.com/dmathieu>
> I got a err message
>
>
> /tmp/build # cd nginx-1.19.10/
> /tmp/build/nginx-1.19.10 # ls
> CHANGES     LICENSE     README      conf        contrib     man         src
> CHANGES.ru  Makefile    auto        configure   html        objs
> /tmp/build/nginx-1.19.10 # ./configure '--prefix=/usr/local/nginx' --with-compat '--add-dynamic-module=/tmp/build/opent
> elemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx'
> checking for OS
>  + Linux 3.10.0-1160.59.1.el7.x86_64 x86_64
> checking for C compiler ... found
>  + using GNU C compiler
>  + gcc version: 10.3.1 20210424 (Alpine 10.3.1_git20210424)
> checking for gcc -pipe switch ... found
> checking for -Wl,-E switch ... found
> checking for gcc builtin atomic operations ... found
> checking for C99 variadic macros ... found
> checking for gcc variadic macros ... found
> checking for gcc builtin 64 bit byteswap ... found
> checking for unistd.h ... found
> checking for inttypes.h ... found
> checking for limits.h ... found
> checking for sys/filio.h ... not found
> checking for sys/param.h ... found
> checking for sys/mount.h ... found
> checking for sys/statvfs.h ... found
> checking for crypt.h ... found
> checking for Linux specific features
> checking for epoll ... found
> checking for EPOLLRDHUP ... found
> checking for EPOLLEXCLUSIVE ... found
> checking for eventfd() ... found
> checking for O_PATH ... found
> checking for sendfile() ... found
> checking for sendfile64() ... found
> checking for sys/prctl.h ... found
> checking for prctl(PR_SET_DUMPABLE) ... found
> checking for prctl(PR_SET_KEEPCAPS) ... found
> checking for capabilities ... not found
> checking for crypt_r() ... found
> checking for sys/vfs.h ... found
> checking for nobody group ... found
> checking for poll() ... found
> checking for /dev/poll ... not found
> checking for kqueue ... not found
> checking for crypt() ... found
> checking for F_READAHEAD ... not found
> checking for posix_fadvise() ... found
> checking for O_DIRECT ... found
> checking for F_NOCACHE ... not found
> checking for directio() ... not found
> checking for statfs() ... found
> checking for statvfs() ... found
> checking for dlopen() ... found
> checking for sched_yield() ... found
> checking for sched_setaffinity() ... found
> checking for SO_SETFIB ... not found
> checking for SO_REUSEPORT ... found
> checking for SO_ACCEPTFILTER ... not found
> checking for SO_BINDANY ... not found
> checking for IP_TRANSPARENT ... found
> checking for IP_BINDANY ... not found
> checking for IP_BIND_ADDRESS_NO_PORT ... found
> checking for IP_RECVDSTADDR ... not found
> checking for IP_SENDSRCADDR ... not found
> checking for IP_PKTINFO ... found
> checking for IPV6_RECVPKTINFO ... found
> checking for TCP_DEFER_ACCEPT ... found
> checking for TCP_KEEPIDLE ... found
> checking for TCP_FASTOPEN ... found
> checking for TCP_INFO ... found
> checking for accept4() ... found
> checking for int size ... 4 bytes
> checking for long size ... 8 bytes
> checking for long long size ... 8 bytes
> checking for void * size ... 8 bytes
> checking for uint32_t ... found
> checking for uint64_t ... found
> checking for sig_atomic_t ... found
> checking for sig_atomic_t size ... 4 bytes
> checking for socklen_t ... found
> checking for in_addr_t ... found
> checking for in_port_t ... found
> checking for rlim_t ... found
> checking for uintptr_t ... uintptr_t found
> checking for system byte ordering ... little endian
> checking for size_t size ... 8 bytes
> checking for off_t size ... 8 bytes
> checking for time_t size ... 8 bytes
> checking for AF_INET6 ... found
> checking for setproctitle() ... not found
> checking for pread() ... found
> checking for pwrite() ... found
> checking for pwritev() ... found
> checking for strerrordesc_np() ... not found
> checking for sys_nerr ... not found
> checking for _sys_nerr ... not found
> checking for localtime_r() ... found
> checking for clock_gettime(CLOCK_MONOTONIC) ... found
> checking for posix_memalign() ... found
> checking for memalign() ... found
> checking for mmap(MAP_ANON|MAP_SHARED) ... found
> checking for mmap("/dev/zero", MAP_SHARED) ... found
> checking for System V shared memory ... found
> checking for POSIX semaphores ... found
> checking for struct msghdr.msg_control ... found
> checking for ioctl(FIONBIO) ... found
> checking for ioctl(FIONREAD) ... found
> checking for struct tm.tm_gmtoff ... found
> checking for struct dirent.d_namlen ... not found
> checking for struct dirent.d_type ... found
> checking for sysconf(_SC_NPROCESSORS_ONLN) ... found
> checking for sysconf(_SC_LEVEL1_DCACHE_LINESIZE) ... not found
> checking for openat(), fstatat() ... found
> checking for getaddrinfo() ... found
> configuring additional dynamic modules
> adding module in /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx
>  + otel_ngx_module was configured
> checking for PCRE library ... found
> checking for PCRE JIT support ... found
> checking for zlib library ... found
> creating objs/Makefile
>
> Configuration summary
>   + using system PCRE library
>   + OpenSSL library is not used
>   + using system zlib library
>
>   nginx path prefix: "/usr/local/nginx"
>   nginx binary file: "/usr/local/nginx/sbin/nginx"
>   nginx modules path: "/usr/local/nginx/modules"
>   nginx configuration prefix: "/usr/local/nginx/conf"
>   nginx configuration file: "/usr/local/nginx/conf/nginx.conf"
>   nginx pid file: "/usr/local/nginx/logs/nginx.pid"
>   nginx error log file: "/usr/local/nginx/logs/error.log"
>   nginx http access log file: "/usr/local/nginx/logs/access.log"
>   nginx http client request body temporary files: "client_body_temp"
>   nginx http proxy temporary files: "proxy_temp"
>   nginx http fastcgi temporary files: "fastcgi_temp"
>   nginx http uwsgi temporary files: "uwsgi_temp"
>   nginx http scgi temporary files: "scgi_temp"
>
> /tmp/build/nginx-1.19.10 # ls -al /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrum
> entation/nginx/
> CMakeLists.txt  README.md       ci/             config          nginx.cmake     src/            test/
>
> /tmp/build/nginx-1.19.10 # ls -al /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrum
> entation/nginx/
> total 24
> drwxrwxr-x    5 root     root           111 Apr  6 07:07 .
> drwxrwxr-x    5 root     root            61 Apr  6 07:07 ..
> -rw-rw-r--    1 root     root          1019 Mar 29 04:20 CMakeLists.txt
> -rw-rw-r--    1 root     root          8090 Mar 29 04:20 README.md
> drwxrwxr-x    2 root     root            34 Apr  6 07:07 ci
> -rw-rw-r--    1 root     root           908 Mar 29 04:20 config
> -rw-rw-r--    1 root     root          1170 Mar 29 04:20 nginx.cmake
> drwxrwxr-x    2 root     root          4096 Apr  6 07:07 src
> drwxrwxr-x    6 root     root           112 Apr  6 07:07 test
>
> /tmp/build/nginx-1.19.10 # make modules
> make -f objs/Makefile modules
> make[1]: Entering directory '/tmp/build/nginx-1.19.10'
> cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
>         -o objs/addon/src/agent_config.o \
>         /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/agent_config.cpp
> cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
>         -o objs/addon/src/nginx_config.o \
>         /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/nginx_config.cpp
> cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
>         -o objs/addon/src/otel_ngx_module.o \
>         /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/otel_ngx_module.cpp
> cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
>         -o objs/addon/src/propagate.o \
>         /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/propagate.cpp
> cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
>         -o objs/addon/src/script.o \
>         /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/script.cpp
> cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
>         -o objs/addon/src/toml.o \
>         /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/toml.c
> cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
>         -o objs/addon/src/trace_context.o \
>         /tmp/build/opentelemetry-cpp-contrib-2656a4072e257b6794da86ddd1b773b49f5517b3/instrumentation/nginx/src/trace_context.cpp
> cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
>         -o objs/otel_ngx_module_modules.o \
>         objs/otel_ngx_module_modules.c
> cc -o objs/otel_ngx_module.so \
> objs/addon/src/agent_config.o \
> objs/addon/src/nginx_config.o \
> objs/addon/src/otel_ngx_module.o \
> objs/addon/src/propagate.o \
> objs/addon/src/script.o \
> objs/addon/src/toml.o \
> objs/addon/src/trace_context.o \
> objs/otel_ngx_module_modules.o \
> -lstdc++ -lopentelemetry_common -lopentelemetry_resources -lopentelemetry_trace -lopentelemetry_exporter_otlp_grpc -lopentelemetry_otlp_recordable \
> -shared
> /usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lopentelemetry_common
> /usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lopentelemetry_resources
> /usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lopentelemetry_trace
> /usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lopentelemetry_exporter_otlp_grpc
> /usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lopentelemetry_otlp_recordable
> collect2: error: ld returned 1 exit status
> make[1]: *** [objs/Makefile:1210: objs/otel_ngx_module.so] Error 1
> make[1]: Leaving directory '/tmp/build/nginx-1.19.10'
> make: *** [Makefile:16: modules] Error 2
>
> Is this a related error?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/ingress-nginx/issues/8437#issuecomment-1089946230>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABGZVWXAIFMN7CUNJQAE5HTVDU7A7ANCNFSM5SU3ENIQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

--- a/images/opentelemetry/rootfs/CMakeLists.txt
+++ b/images/opentelemetry/rootfs/CMakeLists.txt
@@ -1,0 +1,95 @@
+#!/bin/bash
+
+# Copyright 2021 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+cmake_minimum_required(VERSION 3.11 FATAL_ERROR)
+
+project(
+  dependencies
+  LANGUAGES CXX
+  VERSION 0.0.1)
+
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_EXTENSIONS OFF)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_FLAGS "-O2")
+
+set(CMAKE_BUILD_TYPE
+    Release
+    CACHE STRING "Build type" FORCE)
+
+include(GNUInstallDirs)
+
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY
+    ${PROJECT_BINARY_DIR}/${CMAKE_INSTALL_LIBDIR})
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY
+    ${PROJECT_BINARY_DIR}/${CMAKE_INSTALL_LIBDIR})
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY
+    ${PROJECT_BINARY_DIR}/${CMAKE_INSTALL_BINDIR})
+
+set(INSTALL_LIBDIR
+    ${CMAKE_INSTALL_LIBDIR}
+    CACHE PATH "directory for libraries")
+set(INSTALL_BINDIR
+    ${CMAKE_INSTALL_BINDIR}
+    CACHE PATH "directory for executables")
+set(INSTALL_INCLUDEDIR
+    ${CMAKE_INSTALL_INCLUDEDIR}
+    CACHE PATH "directory for header files")
+
+set(DEF_INSTALL_CMAKEDIR share/cmake/${PROJECT_NAME})
+set(INSTALL_CMAKEDIR
+    ${DEF_INSTALL_CMAKEDIR}
+    CACHE PATH "directory for CMake files")
+
+set_property(DIRECTORY PROPERTY EP_BASE ${CMAKE_BINARY_DIR}/subs)
+
+set(STAGED_INSTALL_PREFIX ${CMAKE_BINARY_DIR}/stage)
+message(STATUS "${PROJECT_NAME} staged install: ${STAGED_INSTALL_PREFIX}")
+
+find_package(OpenSSL REQUIRED)
+message("OpenSSL include dir: ${OPENSSL_INCLUDE_DIR}")
+message("OpenSSL libraries: ${OPENSSL_LIBRARIES}")
+
+set(GRPC_GIT_TAG
+    "v1.45.2"
+    CACHE STRING "gRPC version")
+
+include(ExternalProject)
+ExternalProject_Add(
+  gRPC
+  GIT_REPOSITORY https://github.com/grpc/grpc.git
+  GIT_TAG ${GRPC_GIT_TAG}
+  GIT_SHALLOW 1
+  UPDATE_COMMAND ""
+  CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${STAGED_INSTALL_PREFIX}
+             -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+             -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
+             -DgRPC_SSL_PROVIDER=package
+             -DOPENSSL_ROOT_DIR=OpenSSL
+             -DgRPC_BUILD_TESTS=OFF
+             -DBUILD_SHARED_LIBS=ON
+             -DgRPC_INSTALL=ON
+  CMAKE_CACHE_ARGS -DCMAKE_CXX_FLAGS:STRING=${CMAKE_CXX_FLAGS}
+  TEST_AFTER_INSTALL 0
+  DOWNLOAD_NO_PROGRESS 1
+  LOG_CONFIGURE 1
+  LOG_BUILD 0
+  LOG_INSTALL 1)
+
+install(
+  DIRECTORY ${STAGED_INSTALL_PREFIX}/
+  DESTINATION .
+  USE_SOURCE_PERMISSIONS)

--- a/images/opentelemetry/rootfs/Dockerfile
+++ b/images/opentelemetry/rootfs/Dockerfile
@@ -13,16 +13,33 @@
 # limitations under the License.
 
 
-FROM alpine:3.14.6 as builder
+FROM alpine:3.14.6 as base
 
-COPY . /
+RUN mkdir -p /opt/third_party/install
+COPY . /opt/third_party/
 
+# install build tools
 RUN apk update \
 	&& apk upgrade \
 	&& apk add -U bash \
-	&& /build.sh
+	&& bash /opt/third_party/build.sh -p
 
-FROM alpine:3.14.6 
+# install gRPC
+FROM base as grpc
+RUN bash /opt/third_party/build.sh -g v1.43.2
 
-COPY --from=builder init_module.sh /usr/local/bin/init_module.sh
-COPY --from=builder /etc/nginx/modules /etc/nginx/modules
+# install OpenTelemetry-cpp
+FROM base as otel-cpp
+COPY --from=grpc /opt/third_party/install/ /usr
+RUN bash /opt/third_party/build.sh -o v1.3.0
+
+# install otel_ngx_module.so
+FROM base as nginx
+COPY --from=grpc /opt/third_party/install/ /usr
+COPY --from=otel-cpp /opt/third_party/install/ /usr
+RUN bash /opt/third_party/build.sh -n
+
+FROM alpine:3.14.6
+COPY --from=base /opt/third_party/init_module.sh /usr/local/bin/init_module.sh
+COPY --from=nginx /etc/nginx/modules /etc/nginx/modules
+COPY --from=nginx /opt/third_party/install/lib /usr/lib

--- a/images/opentelemetry/rootfs/build.sh
+++ b/images/opentelemetry/rootfs/build.sh
@@ -116,7 +116,7 @@ get_src()
   echo "Downloading $url"
 
   curl -sSL "$url" -o "$f"
-  # echo "$hash  $f" | sha256sum -c - || exit 10
+  echo "$hash  $f" | sha256sum -c - || exit 10
   tar xzf "$f"
   rm -rf "$f"
 }

--- a/images/opentelemetry/rootfs/build.sh
+++ b/images/opentelemetry/rootfs/build.sh
@@ -18,15 +18,12 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-export NGINX_VERSION=1.19.10
-
+export GRPC_GIT_TAG=${GRPC_GIT_TAG:="v1.43.2"}
 # Check for recent changes: https://github.com/open-telemetry/opentelemetry-cpp/compare/v1.2.0...main
-export OPENTELEMETRY_CPP_VERSION=1.2.0
-
-# Check for recent changes: https://github.com/open-telemetry/opentelemetry-cpp-contrib/compare/2656a4...main
-export OPENTELEMETRY_CONTRIB_COMMIT=2656a4072e257b6794da86ddd1b773b49f5517b3
-
-export BUILD_PATH=/tmp/build
+export OPENTELEMETRY_CPP_VERSION=${OPENTELEMETRY_CPP_VERSION:="1.2.0"}
+export INSTAL_DIR=/opt/third_party/install
+# improve compilation times
+CORES=$(($(grep -c ^processor /proc/cpuinfo) - 1))
 
 rm -rf \
    /var/cache/debconf/* \
@@ -35,15 +32,80 @@ rm -rf \
    /tmp/* \
    /var/tmp/*
 
-
-mkdir -p /etc/nginx
+export BUILD_PATH=/tmp/build
 mkdir --verbose -p "$BUILD_PATH"
-cd "$BUILD_PATH"
 
-apk add \
-  curl \
-  git \
-  build-base
+Help()
+{
+   # Display Help
+   echo "Add description of the script functions here."
+   echo
+   echo "Syntax: scriptTemplate [-h|g|o|n|p|]"
+   echo "options:"
+   echo "h     Print Help."
+   echo "g     gRPC git tag"
+   echo "o     OpenTelemetry git tag"
+   echo "n     install nginx"
+   echo "p     prepare"
+   echo
+}
+
+prepare()
+{
+  apk add \
+    linux-headers \
+    openssl \
+    curl-dev \
+    openssl-dev \
+    gtest-dev \
+    c-ares-dev \
+    pcre-dev \
+    curl \
+    git \
+    build-base
+}
+
+install_grpc()
+{
+  mkdir -p $BUILD_PATH/grpc
+  cd ${BUILD_PATH}/grpc
+  cmake -DCMAKE_INSTALL_PREFIX=${INSTAL_DIR} \
+    -DGRPC_GIT_TAG=${GRPC_GIT_TAG} /opt/third_party \
+    -DgRPC_BUILD_GRPC_NODE_PLUGIN=OFF \
+    -DgRPC_BUILD_GRPC_OBJECTIVE_C_PLUGIN=OFF \
+    -DgRPC_BUILD_GRPC_PHP_PLUGIN=OFF \
+    -DgRPC_BUILD_GRPC_PHP_PLUGIN=OFF \
+    -DgRPC_BUILD_GRPC_PYTHON_PLUGIN=OFF \
+    -DgRPC_BUILD_GRPC_RUBY_PLUGIN=OFF
+  cmake --build . -j ${CORES} --target all install
+}
+
+install_otel()
+{
+  cd ${BUILD_PATH}
+  export LD_LIBRARY_PATH="${LD_LIBRARY_PATH:+LD_LIBRARY_PATH:}${INSTAL_DIR}/lib:/usr/local"
+  export PATH="${PATH}:${INSTAL_DIR}/bin"
+  git clone --recurse-submodules -j ${CORES} --depth=1 -b \
+    ${OPENTELEMETRY_CPP_VERSION} https://github.com/open-telemetry/opentelemetry-cpp.git opentelemetry-cpp-${OPENTELEMETRY_CPP_VERSION}
+  cd "opentelemetry-cpp-${OPENTELEMETRY_CPP_VERSION}"
+  mkdir -p .build
+  cd .build
+
+  cmake -DCMAKE_BUILD_TYPE=Release \
+        -DCMAKE_POSITION_INDEPENDENT_CODE=TRUE  \
+        -DWITH_ZIPKIN=OFF \
+        -DWITH_JAEGER=OFF \
+        -DCMAKE_INSTALL_PREFIX=${INSTAL_DIR} \
+        -DBUILD_TESTING=OFF \
+        -DBUILD_SHARED_LIBS=ON \
+        -DWITH_OTLP=ON \
+        -DWITH_OTLP_GRPC=ON \
+        -DWITH_EXAMPLES=OFF \
+        -DWITH_ABSEIL=ON \
+        -DWITH_OTLP_HTTP=OFF \
+        ..
+  cmake --build . -j ${CORES} --target install
+}
 
 get_src()
 {
@@ -54,58 +116,105 @@ get_src()
   echo "Downloading $url"
 
   curl -sSL "$url" -o "$f"
-  echo "$hash  $f" | sha256sum -c - || exit 10
+  # echo "$hash  $f" | sha256sum -c - || exit 10
   tar xzf "$f"
   rm -rf "$f"
 }
 
+install_nginx()
+{
+  export NGINX_VERSION=1.19.10
 
-get_src e8d0290ff561986ad7cd6c33307e12e11b137186c4403a6a5ccdb4914c082d88 \
-        "https://nginx.org/download/nginx-$NGINX_VERSION.tar.gz"
+  # Check for recent changes: https://github.com/open-telemetry/opentelemetry-cpp-contrib/compare/2656a4...main
+  export OPENTELEMETRY_CONTRIB_COMMIT=6467ec2e4d67b08b44580b7eb7a298786f4eef91
 
-get_src 360cdcbd1a235ec62119cc53956b2d31b6ff5f41d44415be53acc544709d58b8 \
-        "https://github.com/open-telemetry/opentelemetry-cpp-contrib/archive/$OPENTELEMETRY_CONTRIB_COMMIT.tar.gz"
+  mkdir -p /etc/nginx
+  cd "$BUILD_PATH"
 
-# improve compilation times
-CORES=$(($(grep -c ^processor /proc/cpuinfo) - 1))
+  get_src 360cdcbd1a235ec62119cc53956b2d31b6ff5f41d44415be53acc544709d58b8 \
+          "https://github.com/open-telemetry/opentelemetry-cpp-contrib/archive/$OPENTELEMETRY_CONTRIB_COMMIT.tar.gz"
 
-export MAKEFLAGS=-j${CORES}
+  cd ${BUILD_PATH}/opentelemetry-cpp-contrib-${OPENTELEMETRY_CONTRIB_COMMIT}/instrumentation/nginx
+  mkdir -p build
+  cd build
+  cmake -DCMAKE_BUILD_TYPE=Release \
+    -DCMAKE_INSTALL_PREFIX=${INSTAL_DIR} \
+    -DBUILD_SHARED_LIBS=ON \
+    -DNGINX_VERSION=${NGINX_VERSION} \
+    ..
+  cmake --build . -j ${CORES} --target install
 
-apk add \
-  protobuf-dev \
-  grpc \
-  grpc-dev \
-  gtest-dev \
-  c-ares-dev \
-  pcre-dev
+  mkdir -p /etc/nginx/modules
+  cp ${INSTAL_DIR}/otel_ngx_module.so /etc/nginx/modules/otel_ngx_module.so
 
-cd $BUILD_PATH
-git clone --recursive https://github.com/open-telemetry/opentelemetry-cpp opentelemetry-cpp-$OPENTELEMETRY_CPP_VERSION
-cd "opentelemetry-cpp-$OPENTELEMETRY_CPP_VERSION"
-git checkout v$OPENTELEMETRY_CPP_VERSION
-mkdir .build
-cd .build
+  mkdir -p ${INSTAL_DIR}/lib
+  cp /usr/lib/libopentelemetry_exporter_otlp_grpc.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libopentelemetry_otlp_recordable.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libprotobuf.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libopentelemetry_trace.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libopentelemetry_resources.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libopentelemetry_common.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libstdc++.so* ${INSTAL_DIR}/lib
 
-cmake -DCMAKE_BUILD_TYPE=Release \
-      -DBUILD_TESTING=OFF \
-      -DWITH_EXAMPLES=OFF \
-      -DCMAKE_POSITION_INDEPENDENT_CODE=ON \
-      -DWITH_OTLP=ON \
-      -DWITH_OTLP_HTTP=OFF \
-      ..
-make
-make install
+  cp /usr/lib/libgrpc.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libgcc_s.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libgrpc++.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_bad_variant_access.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_synchronization.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_raw_hash_set.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_hash.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_statusor.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libgpr.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libupb.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_status.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_time.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_strings.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_stacktrace.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_symbolize.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_malloc_internal.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_base.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_spinlock_wait.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_raw_logging_internal.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libre2.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libaddress_sorting.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_cord.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_bad_optional_access.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_str_format_internal.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_throw_delegate.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_time_zone.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_city.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_low_level_hash.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_cordz_info.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_int128.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_strings_internal.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_debugging_internal.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_cord_internal.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_cordz_functions.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_cordz_handle.so* ${INSTAL_DIR}/lib
+  cp /usr/lib/libabsl_exponential_biased.so* ${INSTAL_DIR}/lib
+}
 
-# build nginx
-cd "$BUILD_PATH/nginx-$NGINX_VERSION"
-./configure \
-  --prefix=/usr/local/nginx \
-  --with-compat \
-  --add-dynamic-module=$BUILD_PATH/opentelemetry-cpp-contrib-$OPENTELEMETRY_CONTRIB_COMMIT/instrumentation/nginx
-
-make modules
-mkdir -p /etc/nginx/modules
-cp objs/otel_ngx_module.so /etc/nginx/modules/otel_ngx_module.so
-
-# remove .a files
-find /usr/local -name "*.a" -print | xargs /bin/rm
+while getopts ":hpng:o:" option; do
+   case $option in
+    h) # display Help
+         Help
+         exit;;
+    g) # install gRPC with git tag
+        GRPC_GIT_TAG=${OPTARG}
+        install_grpc
+        exit;;
+    o) # install OpenTelemetry tag
+        OPENTELEMETRY_CPP_VERSION=${OPTARG}
+        install_otel
+        exit;;
+    p) # prepare
+        prepare
+        exit;;
+    n) # install nginx
+        install_nginx
+        exit;;
+    \?)
+        Help
+        exit;;
+   esac
+done

